hudi-utilities/src/main/java/org/apache/hudi/utilities/sources/helpers/ProtoConversionUtil.java

```diff
@@ -402,7 +438,21 @@ private static Object convertObject(Schema schema, Object value) {
           if (value instanceof Message) {
             // check if this is a Timestamp
             if (LogicalTypes.timestampMicros().equals(schema.getLogicalType())) {
-              return Timestamps.toMicros((Timestamp) value);
+              if (value instanceof Timestamp) {
+                return Timestamps.toMicros((Timestamp) value);
+              } else if (value instanceof DynamicMessage) {
+                Timestamp.Builder builder = Timestamp.newBuilder();
+                ((DynamicMessage) value).getAllFields().forEach((fieldDescriptor, fieldValue) -> {
+                  if (fieldDescriptor.getFullName().equals("google.protobuf.Timestamp.seconds")) {
+                    builder.setSeconds((Long) fieldValue);
+                  } else if (fieldDescriptor.getFullName().equals("google.protobuf.Timestamp.nanos")) {
+                    builder.setNanos((Integer) fieldValue);
+                  }
```

Comment:
Should it be other way around i.e. for Timestamp.seconds, cast to Integer and Timestamp.nanos to Long?

Reply:
Seconds is seconds since epoch and nanos is nanos within the second so the units are as expected